File: server/routes/user_settings.js
import uuid from 'uuid/v4';
import {pick} from 'lodash';
import Joi from 'joi';
import Boom from 'boom';
import UserSettings from '../classes/user_settings';

export default function users(server) {
  server.route({
    /**
     * GET all users settings
     *
     * @header
     *  @param {string} JWT token for authentication
     * @return {array} list of users data
     */
    path: '/api/v3/user/settings',
    method: 'GET',
    config: {
      auth: {
        strategy: 'token',
      },
    },
    handler: async function(request, reply) {
      const settings = new UserSettings({server});

      try {
        const data = await settings.getAll(['guid', 'username', 'partid', 'category', 'param', 'data']);
        if (!data || !data.length) {
          return reply(Boom.notFound('no users settings found'));
        }
        
        data.sort(function(a, b){
              if(a.username < b.username) { return -1; }
              if(a.username > b.username) { return 1; }
              return 0;
        });                                                 
  
        return reply({
          count: data.length,
          data,
        });
      } catch (err) {
        return reply(Boom.serverUnavailable(err));
      }
    },
  });

  server.route({
    /**
     * GET user settings by guid
     *
     * @header
     *  @param {string} JWT token for authentication
     * @request
     *  @param {string} guid
     * @return {array} list of users data
     */
    path: '/api/v3/user/settings/{guid}',
    method: 'GET',
    config: {
      auth: {
        strategy: 'token',
      },
      validate: {
        params: {
          guid: Joi.string().min(12).max(46).required(),
        },
      },
    },
    handler: async function(request, reply) {
      const {guid} = request.params;
      const settings = new UserSettings({server, guid});

      try {
        const data = await settings.get(['guid', 'username', 'partid', 'category', 'param', 'data']);
        if (!data || !Object.keys(data).length) {
          return reply(Boom.notFound('no users settings found for guid ' + guid));
        }
  
        return reply({
          count: data.length,
          data,
        });
      } catch (err) {
        return reply(Boom.serverUnavailable(err));
      }
    },
  });

  server.route({
    /**
     * Create (POST) a new user settings
     *
     * @header
     *  @param {string} JWT token for authentication
     * @payload
     *  @param {string} username
     *  @param {number} partid
     *  @param {string} category
     *  @param {string} param
     *  @param {string} data
     * @return user settings guid and success message
     */
    path: '/api/v3/user/settings',
    method: 'POST',
    config: {
      auth: {
        strategy: 'token',
      },
      validate: {
        payload: {
          username: Joi.string().min(3).max(50).required(),
          partid: Joi.number().required(),
          category: Joi.string().min(3).max(50).required(),
          param: Joi.string().min(3).max(50).required(),
          data: Joi.string(),
        },
      },
    },
    handler: async function(request, reply) {
      const {username, partid, category, param, data} = request.payload;
      const guid = uuid();
      const settings = new UserSettings({server});

      try {
        await settings.add({guid, username, partid, category, param, data});
        return reply({
          data: guid,
          message: 'successfully created user settings',
        }).code(201);
      } catch (err) {
        return reply(Boom.serverUnavailable(err));
      }
    },
  });

  server.route({
    /**
     * Update (PUT) an user setttings
     *
     * @header
     *  @param {string} JWT token for authentication
     * @request
     *  @param {string} guid
     * @payload
     *  @param {string} username
     *  @param {number} partid
     *  @param {string} category
     *  @param {string} param
     *  @param {string} data
     * @return user settings guid and success message
     */
    path: '/api/v3/user/settings/{guid}',
    method: 'PUT',
    config: {
      auth: {
        strategy: 'token',
      },
      validate: {
        params: {
          guid: Joi.string().min(12).max(46).required(),
        },
        payload: {
          username: Joi.string().min(3).max(50),
          partid: Joi.number(),
          category: Joi.string().min(3).max(50),
          param: Joi.string().min(3).max(50),
          data: Joi.string(),
        },
      },
      pre: [
        {
          method: async function(request, reply) {
            const {guid} = request.params;
            const settings = new UserSettings({server, guid});

            try {
              const res = await settings.get(['guid']);
              if (!res || !res.guid || res.guid !== guid) {
                return reply(Boom.notFound(`the user settings with id ${guid} was not found`));
              }

              return reply.continue();
            } catch (err) {
              return reply(Boom.serverUnavailable(err));
            }
          },
        },
      ],
    },
    handler: async function(request, reply) {
      const {guid} = request.params;
      const updates = pick(request.payload, ['username', 'partid', 'category', 'param', 'data']);

      const settings = new UserSettings({server, guid});

      try {
        await settings.update(updates);
        return reply({
          data: guid,
          message: 'successfully updated user settings',
        }).code(201);
      } catch (err) {
        return reply(Boom.serverUnavailable(err));
      }
    },
  });

  server.route({
    /**
     * DELETE an user settings
     *
     * @header
     *  @param {string} JWT token for authentication
     * @request
     *  @param {string} guid
     * @return user settings guid and success message
     */
    path: '/api/v3/user/settings/{guid}',
    method: 'DELETE',
    config: {
      auth: {
        strategy: 'token',
      },
      validate: {
        params: {
          guid: Joi.string().min(12).max(46).required(),
        },
      },
      pre: [
        {
          method: async function(request, reply) {
            const {guid} = request.params;
            const settings = new UserSettings({server, guid});

            try {
              const res = await settings.get(['guid']);
              if (!res || !res.guid || res.guid !== guid) {
                return reply(Boom.notFound(`the user settings with id ${guid} were not found`));
              }

              return reply.continue();
            } catch (err) {
              return reply(Boom.serverUnavailable(err));
            }
          },
        },
      ],
    },
    handler: async function(request, reply) {
      const {guid} = request.params;
      const settings = new UserSettings({server, guid});

      try {
        await settings.delete();
        return reply({
          data: guid,
          message: 'successfully deleted user settings',
        }).code(201);
      } catch (err) {
        return reply(Boom.serverUnavailable(err));
      }
    },
  });
};

File: controller/v1/userSettings.go
package controllerv1

import (
	"fmt"
	"net/http"
	"time"

	"github.com/Jeffail/gabs/v2"
	"github.com/labstack/echo/v4"
	"github.com/sipcapture/homer-app/auth"
	"github.com/sipcapture/homer-app/data/service"
	"github.com/sipcapture/homer-app/model"
	httpresponse "github.com/sipcapture/homer-app/network/response"
	"github.com/sipcapture/homer-app/system/webmessages"
	"github.com/sipcapture/homer-app/utils/logger"
)

type UserSettingsController struct {
	Controller
	UserSettingsService *service.UserSettingsService
}

// swagger:route GET /user/settings settings settingsGetAll
//
// Returns the list of settings
// ---
// produces:
// - application/json
// Security:
// - bearer: []
//
// SecurityDefinitions:
// bearer:
//      type: apiKey
//      name: Authorization
//      in: header
// responses:
//   200: body:UserSettingList
//   400: body:FailureResponse
func (usc *UserSettingsController) GetAll(c echo.Context) error {

	userName, isAdmin := auth.IsRequestAdmin(c)
	reply, err := usc.UserSettingsService.GetAll(userName, isAdmin)
	if err != nil {
		return httpresponse.CreateBadResponse(&c, http.StatusBadRequest, webmessages.UserSettingsFailed)
	}
	return httpresponse.CreateSuccessResponseWithJson(&c, http.StatusOK, []byte(reply))
}

// swagger:route GET /user/settings/{category} settings settingsGetCategory
//
// Returns the list of settings
// ---
// consumes:
// - application/json
// produces:
// - application/json
// parameters:
// + name: category
//   in: path
//   description: user settings category
//   example: dashboard
//   required: true
//   type: string
// Security:
// - bearer: []
//
// SecurityDefinitions:
// bearer:
//      type: apiKey
//      name: Authorization
//      in: header
// responses:
//   200: body:UserSettingList
//   400: body:FailureResponse
func (usc *UserSettingsController) GetCategory(c echo.Context) error {

	userCategory := c.Param("category")
	userName, _ := auth.IsRequestAdmin(c)
	reply, _ := usc.UserSettingsService.GetCategory(userName, userCategory)
	return httpresponse.CreateSuccessResponseWithJson(&c, http.StatusOK, []byte(reply))
}

// swagger:route POST /user/settings settings settingsAddUserSettings
//
// Adds user settings
// ---
// consumes:
// - application/json
// produces:
// - application/json
// parameters:
// + name: UserSetting
//   in: body
//   description: UserSetting struct
//   schema:
//     type: UserSetting
//   required: true
// Security:
// - bearer: []
//
// SecurityDefinitions:
// bearer:
//      type: apiKey
//      name: Authorization
//      in: header
// responses:
//   201: body:UserSettingCreateSuccessfulResponse
//   400: body:FailureResponse
func (usc *UserSettingsController) AddUserSettings(c echo.Context) error {

	userObject := model.TableUserSettings{}
	if err := c.Bind(&userObject); err != nil {
		logger.Error(err.Error())
		return httpresponse.CreateBadResponse(&c, http.StatusBadRequest, webmessages.UserRequestFormatIncorrect)
	}

	if userName, isAdmin := auth.IsRequestAdmin(c); !isAdmin {
		userObject.UserName = userName
	}

	row, _ := usc.UserSettingsService.Add(&userObject)
	return httpresponse.CreateSuccessResponse(&c, http.StatusCreated, row)
}

// swagger:route DELETE /user/settings/{guid} settings settingsDeleteUserSettings
//
// Delete user settings
// ---
// consumes:
// - application/json
// produces:
// - application/json
// parameters:
// + name: guid
//   in: path
//   example: 11111111-1111-1111-1111-111111111111
//   description: guid of user settings
//   required: true
//   type: string
// Security:
// - bearer: []
//
// SecurityDefinitions:
// bearer:
//      type: apiKey
//      name: Authorization
//      in: header
// responses:
//   201: body:UserSettingDeleteSuccessfulResponse
// 	 400: body:FailureResponse
func (usc *UserSettingsController) DeleteUserSettings(c echo.Context) error {

	userObject := model.TableUserSettings{}

	userObject.GUID = c.Param("category")
	userName, isAdmin := auth.IsRequestAdmin(c)

	data, err := usc.UserSettingsService.Get(&userObject, userName, isAdmin)
	if err != nil {
		reply := gabs.New()
		reply.Set(userObject.GUID, "data")
		reply.Set(fmt.Sprintf("the userobject with id %s were not found", userObject.GUID), "message")
	}

	if err := usc.UserSettingsService.Delete(&userObject, userName, isAdmin); err != nil {
		reply := gabs.New()
		reply.Set(userObject.GUID, "data")
		reply.Set(fmt.Sprintf("the userobject with id %s were not found", userObject.GUID), "message")
	}

	reply := gabs.New()
	reply.Set(data.GUID, "data")
	reply.Set("successfully deleted userobject", "message")

	return httpresponse.CreateSuccessResponse(&c, http.StatusCreated, reply.String())
}

// swagger:route PUT /user/settings/{guid} settings settingsUpdateUserSettings
//
// Update user settings
// ---
// consumes:
// - application/json
// produces:
// - application/json
// parameters:
// + name: guid
//   in: path
//   example: 11111111-1111-1111-1111-111111111111
//   description: guid of user settings
//   required: true
//   type: string
// + name: UserSetting
//   in: body
//   description: UserSetting
//   schema:
//     type: UserSetting
//   required: true
// Security:
// - bearer: []
//
// SecurityDefinitions:
// bearer:
//      type: apiKey
//      name: Authorization
//      in: header
// responses:
//   201: body:UserSettingUpdateSuccessfulResponse
// 	 400: body:FailureResponse
func (usc *UserSettingsController) UpdateUserSettings(c echo.Context) error {

	userObject := model.TableUserSettings{}

	if err := c.Bind(&userObject); err != nil {
		logger.Error(err.Error())
		return httpresponse.CreateBadResponse(&c, http.StatusBadRequest, webmessages.UserRequestFormatIncorrect)
	}
	// validate input request body
	if err := c.Validate(userObject); err != nil {
		logger.Error(err.Error())
		return httpresponse.CreateBadResponse(&c, http.StatusBadRequest, err.Error())
	}
	userObject.GUID = c.Param("category")
	userName, isAdmin := auth.IsRequestAdmin(c)

	data, err := usc.UserSettingsService.Get(&userObject, userName, isAdmin)
	if err != nil {
		reply := gabs.New()
		reply.Set(userObject.GUID, "data")
		reply.Set(fmt.Sprintf("the userobject with id %s were not found", userObject.GUID), "message")
	}

	userObject.CreateDate = time.Now()
	userObject.Id = data.Id
	if err := usc.UserSettingsService.Update(&userObject, userName, isAdmin); err != nil {
		reply := gabs.New()
		reply.Set(userObject.GUID, "data")
		reply.Set(fmt.Sprintf("the userobject with id %s were not found", userObject.GUID), "message")
	}

	reply := gabs.New()
	reply.Set(data.GUID, "data")
	reply.Set("successfully updated alias", "message")

	return httpresponse.CreateSuccessResponse(&c, http.StatusCreated, reply.String())
}
